### PR TITLE
Swss.orchagent.multi.consumer

### PR DESCRIPTION
**Pull request by hrachya-mellanox**

### CODE REVIEW

**Comment by stcheng** [comment]
@lguohan @eladraz

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -18,13 +18,13 @@ IntfsOrch::IntfsOrch(DBConnector *db, string tableName, PortsOrch *portsOrch) :
 {
 }
 
-void IntfsOrch::doTask()
+void IntfsOrch::doTask(_in_ Consumer& consumer_info)
```
Comment (stcheng) [inline orchagent/intfsorch.cpp]: change consumer_info to consumer

File: orchagent/neighorch.cpp [hunk 2]
```diff
@@ -5,37 +5,37 @@
 extern sai_neighbor_api_t*         sai_neighbor_api;
 extern sai_next_hop_api_t*         sai_next_hop_api;
 
-void NeighOrch::doTask()
+void NeighOrch::doTask(_in_ Consumer& consumer_info)
```
Comment (stcheng) [inline orchagent/neighorch.cpp]: same here as above.

File: orchagent/neighorch.h [hunk 3]
```diff
@@ -23,14 +23,15 @@ typedef map<NeighborEntry, MacAddress> NeighborTable;
 class NeighOrch : public Orch
 {
 public:
-
     NeighOrch(DBConnector *db, string tableName, PortsOrch *portsOrch, RouteOrch *routeOrch) :
         Orch(db, tableName), m_portsOrch(portsOrch), m_routeOrch(routeOrch) {};
+    const char* const delimiter         = ";";
```
Comment (stcheng) [inline orchagent/neighorch.h]: the delimiter is ':'

File: orchagent/intfsorch.h [hunk 4]
```diff
@@ -16,8 +16,7 @@ class IntfsOrch : public Orch
 public:
     IntfsOrch(DBConnector *db, string tableName, PortsOrch *portsOrch);
 private:
-    void doTask();
-
+    virtual void doTask(_in_ Consumer& consumer_info);
```
Comment (stcheng) [inline orchagent/intfsorch.h]: why it is virtual?

File: orchagent/neighorch.h [hunk 5]
```diff
@@ -23,14 +23,15 @@ typedef map<NeighborEntry, MacAddress> NeighborTable;
...
+    const char* const neigh_field_name  = "neigh";
 private:
     PortsOrch *m_portsOrch;
     RouteOrch *m_routeOrch;
 
-    void doTask();
+    virtual void doTask(_in_ Consumer& consumer_info);
```
Comment (stcheng) [inline orchagent/neighorch.h]: why it is virtual? and i don't think it's necessary to add _in_ here. let's make the code unified everywhere. We are not adding this here. The reason SAI header use _in_ and _out_ is because that is an interface API definition file. These are pure codes and coder shall be able to get the idea easily.

File: orchagent/orch.cpp [hunk 6]
```diff
@@ -1,66 +1,105 @@
...
 #include "logger.h"
 
+using namespace swss;
+
 Orch::Orch(DBConnector *db, string tableName) :
-    m_db(db), m_name(tableName)
+    m_db(db)
 {
-    m_consumer = new ConsumerTable(m_db, tableName);
+    Consumer c_info(new ConsumerTable(m_db, tableName));
```
Comment (stcheng) [inline orchagent/orch.cpp]: change c_info to consumer

File: orchagent/orch.cpp [hunk 7]
```diff
@@ -1,66 +1,105 @@
...
+    m_consumer_map.insert(ConsumerMapPair(tableName, c_info));
+}
+
+Orch::Orch(DBConnector *db, vector<string> &tableNames) :
+    m_db(db)
+{
+    for( auto it = tableNames.begin(); it != tableNames.end(); it++) {
+        Consumer c_info(new ConsumerTable(m_db, *it));
```
Comment (stcheng) [inline orchagent/orch.cpp]: change c_info to consumer

File: orchagent/orch.cpp [hunk 8]
```diff
@@ -1,66 +1,105 @@
...
 {
-    KeyOpFieldsValuesTuple t;
-    m_consumer->pop(t);
+    consumers.clear();
+    for(auto it : m_consumer_map) {
+        consumers.push_back(it.second.m_consumer);
+    }
+}
+ 
+bool Orch::is_owned_consumer(ConsumerTable*consumer) const
```
Comment (stcheng) [inline orchagent/orch.cpp]: space between class name and *

File: orchagent/orch.cpp [hunk 9]
```diff
@@ -1,66 +1,105 @@
...
     }
     /* If an old task is still there, we combine the old task with new task */
     else
     {
-        KeyOpFieldsValuesTuple u = m_toSync[key];
+        KeyOpFieldsValuesTuple existing_data = consumer.m_toSync[key];
 
-        auto tt = kfvFieldsValues(t);
-	auto uu = kfvFieldsValues(u);
+        auto new_values      = kfvFieldsValues(new_data);
```
Comment (stcheng) [inline orchagent/orch.cpp]: we don't need to align = here. just use one space.

File: orchagent/orch.h [hunk 10]
```diff
@@ -15,28 +15,38 @@ extern "C" {
...
     DBConnector *m_db;
-    const string m_name;
+    const string m_name;// TODO: where is this field initialized??
 
 protected:
-    ConsumerTable *m_consumer;
-
-    /* Store the latest 'golden' status */
-    map<string, KeyOpFieldsValuesTuple> m_toSync;
+    ConsumerMap m_consumer_map;
```
Comment (stcheng) [inline orchagent/orch.h]: m_consumerMap

File: orchagent/orch.h [hunk 11]
```diff
@@ -15,28 +15,38 @@ extern "C" {
...
+    bool execute(string tableName);
 
     inline string getOrchName() { return m_name; }
 
+protected:
+    virtual void doTask(_in_ Consumer& consumer_info) = 0;
 private:
     DBConnector *m_db;
-    const string m_name;
+    const string m_name;// TODO: where is this field initialized??
```
Comment (stcheng) [inline orchagent/orch.h]: This value was initialized, but since you're having multiple tables, it will not be valid anymore.

File: orchagent/orch.h [hunk 12]
```diff
@@ -15,28 +15,38 @@ extern "C" {
...
     Orch(DBConnector *db, string tableName);
+    Orch(DBConnector *db, vector<string> &tableNames);
     ~Orch();
 
-    inline ConsumerTable *getConsumer() { return m_consumer; }
+    void getSelectables( _out_ std::vector<Selectable*>& selectables);
+    bool is_owned_consumer(ConsumerTable* s)const;
 
-    void execute();
-    virtual void doTask() = 0;
```
Comment (stcheng) [inline orchagent/orch.h]: Why this is removed.

File: orchagent/orch.h [hunk 13]
```diff
@@ -15,28 +15,38 @@ extern "C" {
...
 class Orch
 {
 public:
     Orch(DBConnector *db, string tableName);
+    Orch(DBConnector *db, vector<string> &tableNames);
     ~Orch();
 
-    inline ConsumerTable *getConsumer() { return m_consumer; }
+    void getSelectables( _out_ std::vector<Selectable*>& selectables);
+    bool is_owned_consumer(ConsumerTable* s)const;
```
Comment (stcheng) [inline orchagent/orch.h]: change function name to hasConsumer

File: orchagent/orch.h [hunk 14]
```diff
@@ -15,28 +15,38 @@ extern "C" {
...
+
 class Orch
 {
 public:
     Orch(DBConnector *db, string tableName);
+    Orch(DBConnector *db, vector<string> &tableNames);
     ~Orch();
 
-    inline ConsumerTable *getConsumer() { return m_consumer; }
+    void getSelectables( _out_ std::vector<Selectable*>& selectables);
```
Comment (stcheng) [inline orchagent/orch.h]: why not return vector<ConsumerTable> and use function name getConsumers

File: orchagent/orchdaemon.cpp [hunk 15]
```diff
@@ -51,27 +73,27 @@ void OrchDaemon::start()
...
         ret = m_select->select(&s, &fd, 1);
         if (ret == Select::ERROR)
-            SWSS_LOG_NOTICE("Error!\n");
+            SWSS_LOG_ERROR("Failed to obtain selectable\n");
 
         if (ret == Select::TIMEOUT)
             continue;
 
-        Orch *o = getOrchByConsumer((ConsumerTable *)s);
+        Orch *o = getOrchByConsumer((ConsumerTable *)s);// NOTE: o can be nullptr
```
Comment (stcheng) [inline orchagent/orchdaemon.cpp]: assert o is not a nullptr?

File: orchagent/orchdaemon.h [hunk 16]
```diff
@@ -10,6 +10,8 @@
 #include "intfsorch.h"
 #include "neighorch.h"
 #include "routeorch.h"
+#include "qosorch.h"
+
```
Comment (stcheng) [inline orchagent/orchdaemon.h]: extra empty line.

File: orchagent/portsorch.cpp [hunk 17]
```diff
@@ -149,13 +149,13 @@ bool PortsOrch::setPortAdminStatus(sai_object_id_t id, bool up)
     return true;
 }
 
-void PortsOrch::doTask()
+void PortsOrch::doTask(_in_ Consumer& consumer_info)
```
Comment (stcheng) [inline orchagent/portsorch.cpp]: consumer_info to consumer

File: orchagent/qosorch.cpp [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+#include "logger.h"
+
+namespace swss {
+
+void QosOrch::doTask(_in_ Consumer& consumer_info)
+{
+}
+
+}
+
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: Can you split the qosorch to a different pull request or at least different commit?

File: orchagent/qosorch.h [hunk 19]
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
Comment (stcheng) [inline orchagent/qosorch.h]: as long as other places are using #ifndef just do the similar thing here to meet code unification

File: orchagent/routeorch.cpp [hunk 20]
```diff
@@ -7,13 +7,13 @@ extern sai_next_hop_group_api_t*        sai_next_hop_group_api;
 
 extern sai_object_id_t gVirtualRouterId;
 
-void RouteOrch::doTask()
+void RouteOrch::doTask(Consumer& consumer_info)
```
Comment (stcheng) [inline orchagent/routeorch.cpp]: same as above

File: orchagent/intfsorch.h (hunk 4 above)
Comment (hrachya-mellanox) [inline orchagent/intfsorch.h]: because base class is virtual. putting here virtual is to make it explicit and avoid any possible confusion.

File: orchagent/orch.cpp (hunk 8 above)
Comment (hrachya-mellanox) [inline orchagent/orch.cpp]: done

File: orchagent/orch.cpp (hunk 9 above)
Comment (hrachya-mellanox) [inline orchagent/orch.cpp]: done

File: orchagent/orch.h (hunk 10 above)
Comment (hrachya-mellanox) [inline orchagent/orch.h]: done

File: orchagent/orch.h (hunk 11 above)
Comment (hrachya-mellanox) [inline orchagent/orch.h]: done

File: orchagent/orch.h (hunk 13 above)
Comment (hrachya-mellanox) [inline orchagent/orch.h]: done

File: orchagent/orchdaemon.cpp (hunk 15 above)
Comment (hrachya-mellanox) [inline orchagent/orchdaemon.cpp]: removed the note

File: orchagent/orchdaemon.h (hunk 16 above)
Comment (hrachya-mellanox) [inline orchagent/orchdaemon.h]: done

File: orchagent/portsorch.cpp (hunk 17 above)
Comment (hrachya-mellanox) [inline orchagent/portsorch.cpp]: done

File: orchagent/qosorch.h (hunk 19 above)
Comment (hrachya-mellanox) [inline orchagent/qosorch.h]: done

File: orchagent/routeorch.cpp (hunk 20 above)
Comment (hrachya-mellanox) [inline orchagent/routeorch.cpp]: done

File: orchagent/neighorch.h (hunk 3 above)
Comment (hrachya-mellanox) [inline orchagent/neighorch.h]: removed, will put in separate commit.

**Comment by lguohan** [comment]
@hrachya-mellanox , can you close this one as you have a new one?
